File: JAVA/_2009_MinimumNumberOfOperationsToMakeArrayContinuous.java
// Arrays, HashSet, Sorting, Sliding Window

// Problem: https://leetcode.com/problems/minimum-number-of-operations-to-make-array-continuous/

import java.util.*;

class _2009_MinimumNumberofOperationstoMakeArrayContinuous {
    public static int minOperations(int[] nums) {
        int maxi = 0;
        int count = 0; 
        int n = nums.length - 1; 
        int l = 0; 

        Arrays.sort(nums); 

        for (int i = 0; i < nums.length; i++) {
            if (i + 1 < nums.length && nums[i] == nums[i + 1]) {
                continue;
            }
            nums[l++] = nums[i];
        }

        for (int i = 0, j = 0; i < l; i++) {
            while (j < l && (nums[j] - nums[i]) <= n) {
                count++;
                j++;
            }
            maxi = Math.max(maxi, count);
            count--;
        }

        return nums.length - maxi;
    }    
    public static void main(String[] args) {
        int[] nums = {1, 10, 100, 1000};
        System.out.println(minOperations(nums));
    }
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: C++/_2009_MinimumNumberOfOperationsToMakeArrayContinuous.cpp
// Arrays, HashSet, Sorting, Sliding Window

// Problem: https://leetcode.com/problems/minimum-number-of-operations-to-make-array-continuous/

#include <iostream>
#include <bits/stdc++.h>
#include <vector>
using namespace std;

class Solution{
public:
    static int minOperations(vector<int>& nums){
        int n = nums.size();
        unordered_set<int> s;
        for(int i = 0; i < n; i++){
            s.insert(nums[i]);
        }
        vector<int> a;
        for(auto i : s){
            a.push_back(i);
        }
        sort(a.begin(), a.end());
        int res = n;
        int r = 0;

        for(int l = 0; l < a.size(); l++){
            while(r < a.size() && a[r] < a[l] + n){
                r += 1;
            }
            int window = r - l;
            res = min(res, n - window);
        }

        return res;
    }
};

int main(){
    vector<int> nums = {4, 2, 5, 3};
    cout << Solution::minOperations(nums) << endl;
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)